src/box/lua/tuple.c

```diff
@@ -749,18 +748,143 @@ lbox_tuple_format_gc(struct lua_State *L)
+	struct mpstream stream;
+	mpstream_init(&stream, region, region_reserve_cb, region_alloc_cb,
+		      luamp_error, L);
+	mpstream_encode_tuple_field(&stream, fmt, field_no, true);
+	mpstream_flush(&stream);
+	size_t size = region_used(region) - region_svp;
+	const char *data = xregion_join(region, size);
+	luamp_decode(L, luaL_msgpack_default, &data);
+	region_truncate(region, region_svp);
+}
+
+/*
+ * Next function of `{i}pairs`.
+ */
+static int
+luaT_tuple_format_iterator_next(struct lua_State *L)
+{
+	struct tuple_format *fmt = luaT_check_tuple_format(L, 1);
+	uint32_t field_cnt = tuple_format_field_count(fmt);
+	if (field_cnt == 0) {
+		lua_pushnil(L);
+		return 1;
+	}
+
+	if (lua_isnil(L, 2)) {
+		lua_pushnumber(L, 1);
+		push_tuple_field_format(L, fmt, 0);
+		return 2;
+	}
+
+	uint32_t i = luaL_touint64(L, 2);
+	if (i >= field_cnt) {
+		lua_pushnil(L);
+		return 1;
+	}
+	lua_pushnumber(L, i + 1);
+	push_tuple_field_format(L, fmt, i);
+	return 2;
+}
+
+/*
+ * Returns (i, {name, type}) pairs.
+ */
+static int
+lbox_tuple_format_pairs(struct lua_State *L)
+{
+	luaT_check_tuple_format(L, 1);
+	lua_rawgeti(L, LUA_REGISTRYINDEX, luaT_tuple_format_iterator_next_ref);
+	lua_insert(L, 1);
+	lua_pushnil(L);
+	return 3;
+}
+
+/*
+ * Returns an array tuple field formats.
+ */
+static int
+lbox_tuple_format_serialize(struct lua_State *L)
+{
+	struct tuple_format *fmt = luaT_check_tuple_format(L, 1);
+	const char *fmt_mp = fmt->mp;
+	luamp_decode(L, luaL_msgpack_default, &fmt_mp);
+	luaL_findtable(L, LUA_GLOBALSINDEX, "box.internal.space", 1);
+	lua_getfield(L, -1, "denormalize_format");
+	lua_remove(L, -2);
+	lua_pushvalue(L, -2);
+	lua_call(L, 1, 1);
+	return 1;
+}
+
 int
 lbox_push_tuple_format(struct lua_State *L, struct tuple_format *format)
 {
-	struct tuple_format **ptr = (struct tuple_format **)
-		luaL_pushcdata(L, CTID_STRUCT_TUPLE_FORMAT_PTR);
+	struct tuple_format **ptr = lua_newuserdata(L, sizeof(*ptr));
 	*ptr = format;
 	tuple_format_ref(format);
-	lua_pushcfunction(L, lbox_tuple_format_gc);
-	luaL_setcdatagc(L, -2);
+	luaL_getmetatable(L, tuple_formatlib_name);
+	lua_setmetatable(L, -2);
 	return 1;
 }
 
+static int
+lbox_tuple_format_new(struct lua_State *L)
+{
+	int top = lua_gettop(L);
+	if (top == 0)
+		return lbox_push_tuple_format(L, tuple_format_runtime);
+	assert(top == 1 && lua_istable(L, 1));
+	uint32_t count = lua_objlen(L, 1);
+	if (count == 0)
+		return lbox_push_tuple_format(L, tuple_format_runtime);
+	struct ibuf *buf = cord_ibuf_take();
+	size_t unused;
+	const char *fmt = luaT_tuple_encode_on_lua_ibuf(L, 1, &unused, buf);
+	if (fmt == NULL) {
+		cord_ibuf_drop(buf);
+		luaT_error(L);
+	}
+	struct region *region = &fiber()->gc;
+	size_t region_svp = region_used(region);
+	struct field_def *fields;
+	uint32_t field_count;
+	int rc = field_def_array_decode(&fmt, &fields, &field_count, region);
+	cord_ibuf_drop(buf);
+	if (rc != 0) {
+		region_truncate(region, region_svp);
+		return luaT_error(L);
+	}
```

Comment:
Is it a kind of 'get all `name` fields from an array'? Why it is done using encoding to msgpack and decoding back?

Or it also validates the incoming data?

Wouldn't it be more obvious if we would implement a direct construction of a format from Lua, just like in `lbox_key_def_new()` + `luaT_key_def_set_part()`?

Reply:
We wanted to reuse `field_def_array_decode`.

> Wouldn't it be more obvious if we would implement a direct construction of a format from Lua, just like in `lbox_key_def_new()` + `luaT_key_def_set_part()`?

`luaT_key_def_set_part()` is 100 lines of code, and it supports less than half of the options available in format clause, imagine how much more lines would it take to support all of them — and these functions are definitely not performance critical.

Reply:
At least some explanation what does it do would be nice. As I understood, something like the following.

> The format definition is accepted as a Lua table. It has the following shape.
>
> format_def := array-of(field_def)
> field_def := <...> (I expect that the field of this table will be at least listed here)
>
> The definition is encoded into msgpack to reuse existing field_def constructor.
>
> The field_def_array_decode() function performs validation and creates an array of field definitions.

Reply:
Added a comment to `lbox_tuple_format_new`.